src/python/pants/backend/python/tasks2/resolve_requirements_task_base.py

```diff
@@ -47,10 +47,13 @@ def resolve_requirements(self, req_libs):
       # to cover the empty case.
       if not os.path.isdir(path):
         with safe_concurrent_creation(path) as safe_path:
-          self._build_requirements_pex(interpreter, safe_path, req_libs)
+          self._build_requirements_pex(interpreter, safe_path, req_libs,
+                                       python_dist_targets=python_dist_targets)
     return PEX(path, interpreter=interpreter)
 
-  def _build_requirements_pex(self, interpreter, path, req_libs):
+  def _build_requirements_pex(self, interpreter, path, req_libs, python_dist_targets=None):
     builder = PEXBuilder(path=path, interpreter=interpreter, copy=True)
     dump_requirements(builder, interpreter, req_libs, self.context.log)
+    if python_dist_targets:
+      dump_python_distributions(builder, python_dist_targets, self.workdir, self.context.log)
```

Comment:
what happens here if the py_dist target happens to conflict or overwrite a resolved requirement?

seems to me like we should be dumping those prior to the resolve and then considering them in the full transitive resolve closure to reveal any resolution impact.

this also raises the question of how to handle transitive requirements of `python_dist` setup.py.. ideally we could construct the wheel intransitively and then let the resolve handle pulling in the transitive requirements as part of this.

there's also a question of how to handle `setup_requires` in a given `setup.py`, but that's something we can probably just punt on for now and revisit.

Reply:
That's a good point. Would this approach require anything beyond moving this logic up higher (above dump_requirements) and letting `dump_requirements` de-dupe transitive requirements?

How would we ensure that 3rd party requirements resolved as part of a python distribution (specified by `3rdparty/python:dep`) and requirements specified in setup.py are not duplicated? It seems like this would necessitate handling `setup_requires` in the here and now as well.